### ISSUE TITLE
Adding a operation inside entity

### ISSUE THREAD

**Issue by AnderssonPeter** [EntityGraphQL/EntityGraphQL#61]
Lets say i have the following:

```graphql
type RootQuery {
	"Return a Device by its Id"
	device(id: Int!): Device
	devices: [Device!]
}

type Device {
	id: Int!
        name: String
	logs: [LogEntry!]
}

type LogEntry {
	id: Int!
	when: Date!
        isError: Boolean!
	message: String
}
```

Is there some way to add a operation the `Device` type so that can get all logs that match a EQL expression filter? (i don't know if that is valid GraphQL or if its already supported by this package)

This would allow us to change the schema to:
```graphql
type Device {
	id: Int!
        name: String
	logs: [LogEntry!]
        filteredLogs(filter: String!): [LogEntry!]
}
```

This would allow us to execute the following query to only get errors
```graphql
devices {
  id,
  name,
  filteredLogs(filter: 'isError') {
    when,
    message
  }
}
```

**Comment by lukemurray** [comment]
A couple of ways you can do this.

```c#
// assume you have this somewhere
var schema = new SchemaProvider<MyContext>();

// assuming Device is a type in the MyContext object graph and has been automatically added to the schema and we just want to extend it

// the way you described
schema.Type<Device>().AddField("filteredLogs",
   new {
      filter = ArgumentHelper.Required<string>()
   },
   (device, args) => device.Logs.Where(log => log.Type == args.filter),
   "Return device logs filtered by type");
```
You can use `filter =(string)null` to make the argument optional and use the `WhereWhen()` helper to apply the filter when it is set.

You could also use an Enum instead of a string.

**Comment by AnderssonPeter** [comment]
Thanks that worked as intended, i tried to combine it with the `EntityQueryCompiler` but I'm having 2 issues.
1. When inside the `addField` how do i get access to the claims?
2. The example in the documentation yields some error when i try to run `LogEntries.Where(result.LambdaExpression)` i get the error `cannot convert from 'System.Linq.Expressions.LambdaExpression' to 'System.Linq.Expressions.Expression<System.Func<MiFloraGateway.Database.LogEntry, bool>>'`.

**Comment by lukemurray** [comment]
Hi, You mean the `EntityQuery<T>` argument type?

```c#
schema.Type<Device>().AddField("filteredLogs",
   new {
      filter = ArgumentHelper.EntityQuery<LogEntry>()
   },
   (device, args) => device.Logs.WhereWhen(p.filter, p.filter.HasValue),
   "Return device logs filtered by expression");
```

Do you require that the filter is always passed in? Wrapping that in a `ArgumentHelper.Required` should work.

If just using `.Where()` I think the compiler is getting confused. You can tell it what you mean by doing `LinqExtensions.Where(device.Logs, p.filter)`.

I'm not sure I follow why you want to access the claims? The claims you pass in while executing the whole query are passed down to the compiler that handles `EntityQuery<LogEntry>`.

**Comment by AnderssonPeter** [comment]
Sorry for the late response, it seems to work as intended!

Sorry but i have a few more questions:

1. But one question how do i filter with dates with and without time? (Tried the following 2020-01-01, "2020-01-01", 2020-01-01 00:00:00 and "2020-01-01 00:00:00" all seem to fail as they are identified as string or UInt64 or String) (Never created anything with `antlr4` if its the `EntityGraphQL.g4` file that is used it does not seem to contain a definition for DateTimes.

2. When querying a entry with a field that has a enum, it seems to return the number of the enum is this intended? (I have no idea)

3. Related to the one above, how do i use enums in `EQL` filters? (If i just try = 0, i get a error that it can't do a binary operation with the enum and UInt64, if i try = Value i just get that the type does not have a field Value).

4. One thing that isn't a problem for me but i'm curios, if i want to block filtering on some fields how would i do that?
My idea was to create a new `SchemaBuilder` and remove the fields i don't want to filtered on and using the `EntityQueryCompiler`, i found no way to use the output with `Where`.

I could setup a wiki page for the project with all the answers you provide if your interested? (or just provide a more complicated example that could also work)

**Comment by lukemurray** [comment]
1. you should be able to define the property in .NET as a `DateTime` and pass a valid formatted string like `"2020-01-01T00:00:00"`. We use this all the time so if it's not working can you provide any more details or get a unit test failing?

2. use `[JsonConverter(typeof(StringEnumConverter))]` on the enum. This is because JSON.NET defaults to the number

3. EQL filters might not support enums. I'll have a look when I get a chance

4. Interesting, you idea could work.You'll need to use the `Where()` extension from `LinqExtensions` in this project.

And I welcome any help with PRs or documentation help :)

**Comment by AnderssonPeter** [comment]
1. I created the following test (and they all fail? I'm not saying all of them should pass but atleast one of them should?, i can PR this unit test but I'm not sure if all those input values are appropriate?)
```
[Theory]
[InlineData("2020-08-11T00:00:00")]
[InlineData("\"2020-08-11T00:00:00\"")]
[InlineData("'2020-08-11T00:00:00'")]
[InlineData("2020-08-11 00:00:00")]
[InlineData("\"2020-08-11 00:00:00\"")]
[InlineData("'2020-08-11 00:00:00'")]
[InlineData("2020-08-11")]
[InlineData("\"2020-08-11\"")]
[InlineData("'2020-08-11'")]
public void TestLinqQueryWorksWithDates(string dateValue)
{
    var schemaProvider = SchemaBuilder.FromObject<Entry>();
    schemaProvider.AddType<DateTime>("DateTime"); //<-- Tried with and without
    var compiledResult = EntityQueryCompiler.Compile($"when >= {dateValue}", schemaProvider, null);
    var list = new List<Entry> {
        new Entry("First") {
            When = new DateTime(2020, 08, 10)
        },
        new Entry("Second") {
            When = new DateTime(2020, 08, 11)
        },
        new Entry("Third") {
            When = new DateTime(2020, 08, 12)
        }
    };
    Assert.Equal(3, list.Count());
    var results = list.Where((Func<Entry, bool>)compiledResult.LambdaExpression.Compile());

    Assert.Equal(2, results.Count());
    Assert.Equal("Second", results.ElementAt(0).Message);
    Assert.Equal("Third", results.ElementAt(1).Message);
}

private class Entry
{
    public Entry(string message)
    {
        Message = message;
    }
    public DateTime When { get; set; }
    public string Message { get; set; }
}
```
2. Using `.AddNewtonsoftJson(options => options.SerializerSettings.Converters.Add(new StringEnumConverter(new CamelCaseNamingStrategy())));` seems to solve this for all enums!
3. :+1:
4. Not 100% sure i get how to convert the `CompiledQueryResult` into a `EntityQueryType`

**Comment by lukemurray** [comment]
Thanks.

1. I though you meant in GQL mode e.g. dates as field or mutation arguments. I have added support for dates and times in `EntityQueryCompiler` in master

2. thanks, I didn't know about the global option!

4. You should also be able to use the claims support. You can mark certain fields that require a certain claim/role and if the user doesn't have that claim/role it will not let them reference those fields.

If this is a public API that you just want to control how users can filters. You could do it the more "GraphQL way" and add a bunch of field arguments for each field you'd like to be able to filter.

If you want to do want you suggested above. Can you not create that new schema with only the fields you want accessible and then just use 
```c#
var exp = EntityQueryCompiler.Compile(filter, mySchema, null);

var result = exp.Execute(contextValue);
```

Expanding on your original idea you might be able to cast the `EntityQueryCompiler` result property `LambdaExpression` to `Expression<Func<TContext, bool>>` and pass that into `Where()`.

If you can provide a working (or failing in this case) example for me that would help.

Thanks

**Comment by lukemurray** [comment]
3. master also has support for enums in `EntityQueryCompiler`

**Comment by AnderssonPeter** [comment]
1. Thanks now it seems to parse dates correctly
2. Happy to help :)
3. Thanks now it seems to parse the enum correctly!
4. If i add a claim then the user would no longer be able to select the fields?
But to add some context, lets say you have a fairly large table (number of rows) and only have indexes on a few fields, then i would be nice if you could limit the filter query to only operate on the fields that have indexes on them.


But now i have a new Problem, it seems EF has some problems when WhenWhere is applied inside a subselect.
i have the following:
```
schema.Type<Device>().AddField("filteredLogs", new
{
    filter = ArgumentHelper.EntityQuery<LogEntry>()
},
(device, args) => device..WhereWhen(args.filter, args.filter.HasValue),
"Return device logs filtered");
```
And i execute the following query:
```
query test1 {
  device(id: 1) {
    filteredLogs(filter: "when > \"2020-01-01\"") {
      when,
      event,
      result      
    }
  }
}
```

But when trying to use it i get the following error:
```
Processing of the LINQ expression '(MaterializeCollectionNavigation(
    navigation: Navigation: Device.Logs,
    subquery: (NavigationExpansionExpression
        Source: DbSet<LogEntry>
            .Where(l => EF.Property<Nullable<int>>(d, \"Id\") != null && EF.Property<Nullable<int>>(d, \"Id\") == EF.Property<Nullable<int>>(l, \"DeviceId\"))
        PendingSelector: l => (NavigationTreeExpression
            Value: (EntityReference: LogEntry)
            Expression: l)
    )
        .Where(i => EF.Property<Nullable<int>>((NavigationTreeExpression
            Value: (EntityReference: Device)
            Expression: d), \"Id\") != null && EF.Property<Nullable<int>>((NavigationTreeExpression
            Value: (EntityReference: Device)
            Expression: d), \"Id\") == EF.Property<Nullable<int>>(i, \"DeviceId\")))
    .WhereWhen(
        source: q_LogEntry => q_LogEntry.Event == Scan, 
        wherePredicate: __Value_filter_HasValue_1)
    .Select(p_LogEntry => new c742d07f-2fc7-4601-ae38-0b894ed2776d{ 
        when = p_LogEntry.When, 
        event = p_LogEntry.Event, 
        result = p_LogEntry.Result 
    }
    )' by 'NavigationExpandingExpressionVisitor' failed. This may indicate either a bug or a limitation in EF Core. See https://go.microsoft.com/fwlink/?linkid=2101433 for more detailed information.
```

So my idea was to change from `WhereWhen` to `Where` but that yields the following error `When called from 'VisitLambda', rewriting a node of type 'System.Linq.Expressions.ParameterExpression' must return a non-null value of the same type. Alternatively, override 'VisitLambda' and change it to not visit children of this type.`

I'm no expert at compile time expressions but the following looks right to me:
![image](https://user-images.githubusercontent.com/1308309/93806585-a3d9e600-fc49-11ea-94ef-4ad48c62f93c.png)
![image](https://user-images.githubusercontent.com/1308309/93806920-2fec0d80-fc4a-11ea-9240-2323a5eaeee6.png)


What confuses me is that the following works:
```
query test1 {
  device(id: 1) {
    id,
    logsAfter(after: "2020-01-01") {
      when,
      event,
      result      
    }
  }
}
```
```
schema.Type<Device>().AddField("logsAfter", new
{
    after = ArgumentHelper.Required<DateTime>()
},
(device, args) => device.Logs.Where(log => log.When >= args.after),
"Return device logs filtered by time");
```

**Comment by lukemurray** [comment]
Can you try changing the `WhenWhere` to this `device.WhereWhen(args.filter.Value, args.filter.HasValue)`. Using the `Value` not the nullable object.

I understand you use case now. Currently that isn't supported. The best way I think to enable this would be to have the `ArgumentHelper.EntityQuery` take an optional schema that will be used for building the filter.

**Comment by AnderssonPeter** [comment]
It does not look like `EntityQueryType<TType>` has a `Value` property i tried with `Query` but it yields the same error
```
Processing of the LINQ expression '(MaterializeCollectionNavigation(
    navigation: Navigation: Device.Logs,
    subquery: (NavigationExpansionExpression
        Source: DbSet<LogEntry>
            .Where(l => EF.Property<Nullable<int>>(d, \"Id\") != null && EF.Property<Nullable<int>>(d, \"Id\") == EF.Property<Nullable<int>>(l, \"DeviceId\"))
        PendingSelector: l => (NavigationTreeExpression
            Value: (EntityReference: LogEntry)
            Expression: l)
    )
        .Where(i => EF.Property<Nullable<int>>((NavigationTreeExpression
            Value: (EntityReference: Device)
            Expression: d), \"Id\") != null && EF.Property<Nullable<int>>((NavigationTreeExpression
            Value: (EntityReference: Device)
            Expression: d), \"Id\") == EF.Property<Nullable<int>>(i, \"DeviceId\")))
    .WhereWhen(
        source: q_LogEntry => q_LogEntry.When > 2020-01-01 00:00:00, 
        wherePredicate: __Value_filter_HasValue_1)
    .Select(p_LogEntry => new c17ae8f0-d846-41b4-b8c2-6f71cff5913a{ 
        when = p_LogEntry.When, 
        event = p_LogEntry.Event, 
        result = p_LogEntry.Result 
    }
    )' by 'NavigationExpandingExpressionVisitor' failed. This may indicate either a bug or a limitation in EF Core. See https://go.microsoft.com/fwlink/?linkid=2101433 for more detailed information.
```

Sadly i get compilation errors if i try to put the `Query` in a normal `Where` if i convert the expression to a `Func<TType, bool>` using `Expression.Lambda<Func<TType, bool>>(Query).Compile()` but that gave me other errors.

**Comment by lukemurray** [comment]
Sorry I did mean `.Query` not `.Value`.

Hmm weird. I use `EntityQuery<T>()` with no issues with EF Core.

Any chance you can make a PR with a test that replicates this error? Or have some sample project I can get?

**Comment by AnderssonPeter** [comment]
I have modified the demo project in the following: https://github.com/AnderssonPeter/EntityGraphQL/tree/SubSelectError
Run the following query to make it 💥 
```
{
  person(id: 2) {
    id,
    firstName,
    lastName,
    filteredDirectorOf(filter: "id = 1") {
      name
    }
  }
}
```

**Comment by AnderssonPeter** [comment]
I did some unrelated searching and found `IMethodCallTranslator` might allow us to use `WhereWhen` inside a sub select.

**Comment by lukemurray** [comment]
@AnderssonPeter have you used `IMethodCallTranslator` before? I have implemented it and it registers but EF never calls the translate method. Not really sure what is going on TBH.

**Comment by AnderssonPeter** [comment]
I haven't used it yet but I'm willing to give it a try :) do you have a branch for it?

**Comment by lukemurray** [comment]
`feature/efcore31`

You'll see `EntityGraphQLDbContextOptionsExtension.ApplyServices` is called on startup which registers `EntityGraphQLMethodCallTranslatorPlugin` which itself registers `WhereWhenMethodCallTranslator` but `WhereWhenMethodCallTranslator.Translate` is never called. 

You can run that above query and still get the error.

**Comment by lukemurray** [comment]
Whoops `IMethodCallTranslatorPlugin` was 2.2. I have switched to `IMethodCallTranslatorProvider` but still no luck

**Comment by lukemurray** [comment]
I believe the core issue here is resolved in 1.0.x with it's two step expression selection - core context data first (EF supported) then he services on top. And can confirm that the `WhereWhen`s are working with EF 3.1+
